### PR TITLE
allow gauge stored in metric registry to be updated

### PR DESCRIPTION
**Pull request by zhtaoxiang**
Currently, if a gauge is created and save into a metric registry, it cannot be updated. This PR makes it changable.

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/metrics/PinotGauge.java [hunk 1]
```diff
@@ -36,10 +38,26 @@
...
+   */
+  default void setValue(T value) {
+    throw new RuntimeException("setValue is not implemented");
+  }
+
+  /**
+   * Sets the gauge value supplier.
+   * @param valueSupplier the gauge value supplier to set.
+   */
+  default void setValueSupplier(Supplier<T> valueSupplier) {
```
Comment (jackjlli) [inline pinot-spi/src/main/java/org/apache/pinot/spi/metrics/PinotGauge.java:53]: I wonder if it's possible to introduce a new interface that specifies the both setValue() and setValueSupplier() methods, and let this PinotGauge interface also extends the new interface you just created?

File: pinot-plugins/pinot-metrics/pinot-yammer/src/main/java/org/apache/pinot/plugin/metrics/yammer/YammerSettableGauge.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.yammer.metrics.core.Gauge;
+import java.util.function.Supplier;
+
+/**
+ * YammerSettableGauge extends {@link Gauge}, allowing setting a value supplier to provide the gauge value
+ *
+ * @param <T> the type of the metric's value
+ */
+public abstract class YammerSettableGauge<T> extends Gauge<T> {
```
Comment (jackjlli) [inline pinot-plugins/pinot-metrics/pinot-yammer/src/main/java/org/apache/pinot/plugin/metrics/yammer/YammerSettableGauge.java:29]: Same here. We may not need an abstract class but more of an interface to specify the following two methods.

File: pinot-spi/src/main/java/org/apache/pinot/spi/metrics/PinotGauge.java (hunk 1 above)
Comment (zhtaoxiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/metrics/PinotGauge.java:47]: that's a good suggestion! updated the code.

**Review by snleee — APPROVED** [review]
LGTM. Thank you for adding this!

**Pull request merged**